Objects/weakrefobject.c

```diff
@@ -776,104 +920,40 @@ _PyWeakref_CallableProxyType = {
     proxy_iternext,                     /* tp_iternext */
 };
 
-
+static PyWeakReference *
+allocate_ref(PyTypeObject *type, PyObject *obj, PyObject *callback)
+{
+    return new_weakref(obj, callback);
+}
 
 PyObject *
 PyWeakref_NewRef(PyObject *ob, PyObject *callback)
 {
-    PyWeakReference *result = NULL;
-    PyWeakReference **list;
-    PyWeakReference *ref, *proxy;
+    return (PyObject *)get_or_create_weakref(&_PyWeakref_RefType, allocate_ref,
+                                             ob, callback);
+}
 
-    if (!_PyType_SUPPORTS_WEAKREFS(Py_TYPE(ob))) {
-        PyErr_Format(PyExc_TypeError,
-                     "cannot create weak reference to '%s' object",
-                     Py_TYPE(ob)->tp_name);
+static PyWeakReference *
+allocate_proxy(PyTypeObject *type, PyObject *obj, PyObject *callback)
+{
+    PyWeakReference *result = new_weakref(obj, callback);
+    if (result == NULL) {
         return NULL;
     }
-    list = GET_WEAKREFS_LISTPTR(ob);
-    get_basic_refs(*list, &ref, &proxy);
-    if (callback == Py_None)
-        callback = NULL;
-    if (callback == NULL)
-        /* return existing weak reference if it exists */
-        result = ref;
-    if (result != NULL)
-        Py_INCREF(result);
+    if (PyCallable_Check(obj)) {
+        Py_SET_TYPE(result, &_PyWeakref_CallableProxyType);
+    }
```

Comment:
This is icky. Can we refactor this so that:

1) The type determination is pulled up into `PyWeakref_NewProxy`
2) `new_weakref` takes the type so that it doesn't have to be changed after creation